plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -120,34 +121,32 @@ computeArity = \case
 
 -- | Given the arity of a function, and the list of arguments applied to it, return whether it is
 -- fully applied or not.
-isFullyApplied :: Arity -> [Arg tyname name uni fun ann] -> Bool
+isFullyApplied :: Arity -> AppContext tyname name uni fun ann -> Bool
 isFullyApplied [] _ = True -- The function needs no more arguments
-isFullyApplied (_lam:_lams) [] = False -- under-application
-isFullyApplied (hdLams:tlLams) (hdArg:tlArg) =
-  case (hdLams, hdArg) of
-    (TermParam, TermArg _) -> isFullyApplied tlLams tlArg
-    (TypeParam, TypeArg _) -> isFullyApplied tlLams tlArg
-    _ ->
-      -- wrong argument type, i.e., we have an ill-typed term here. It's not what we define as fully
-      -- applied. Although if the term was ill-typed before, it will be ill-typed after the
-      -- inlining, and it won't make it any worse, so we could consider accepting this.
-      False
+isFullyApplied (_lam:_lams) AppContextEnd = False -- under-application
+isFullyApplied (TermParam:tlLams) (TermAppContext _ _ ctx) = isFullyApplied tlLams ctx
+isFullyApplied (TypeParam:tlLams) (TypeAppContext _ _ ctx) = isFullyApplied tlLams ctx
+isFullyApplied _ _ =
+    -- wrong argument type, i.e., we have an ill-typed term here. It's not what we define as fully
+    -- applied. Although if the term was ill-typed before, it will be ill-typed after the
+    -- inlining, and it won't make it any worse, so we could consider accepting this.
+    False
```

Comment:
Can't you just do instead `arity == lengthContext`?

Reply:
Yes, but this is a bit more explicit about the weird case. We discussed this a bit before. Maybe it would be simpler and we should just use that.